adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -38,488 +35,387 @@ import scala.reflect.ClassTag
 sealed abstract class ShuffleRegionJoin[T: ClassTag, U: ClassTag, RT, RU]
     extends RegionJoin[T, U, RT, RU] {
 
-  protected val sd: SequenceDictionary
-  protected val partitionSize: Long
-  protected val sc: SparkContext
+  // stores join candidate tuples from right
+  protected val cache: ListBuffer[(ReferenceRegion, U)] = ListBuffer.empty
```

Comment:
I am of the strong opinion that we don't need these abstractions and I'd like to make a case for not rolling back. 

All we are doing in `shuffleRegionjoin` is a single `mapPartitions` (`sweep`) over an `RDD`. There is no need to create a new object for each partition when all we need is an implementation for how to handle the join. We really don't need an object that extends `Iterator` to have all the implementation details for how to handle each respective join. The extra fields and data structures in that object live with the `Iterator`, even though they aren't needed anymore. In this case, I would prefer to have objects that control how the transformations occur rather than storing the data (in addition to all the fields and transformations required).

There are other reasons to not roll back related to other set-theory operations following the same `sweep` paradigm. Truthfully, I wanted to push the logic up even higher, but that wasn't necessary for this PR.

I share your concerns about the `cache`, though _as far as I can tell from testing_ it is operating as if each thread has its own copy. It wouldn't take much effort to move that into the `sweep` call though and not have to worry about thread safety.